harness/ff-proxy
# FFM-5474 Flush metrics to Saas based on timer or size in MB

**What**

Makes a change so that when operating as a Primary the Proxy flushes metrics to Saas every minute or if the amount of metrics data it's holding in memory exceeds 1MB.

In order to do this I made a fairly substantial refactor to the `metricsservice` package. Before this refactor all the logic for aggregating metrics data by environment and forwarding it on to Saas existed in the `client.go`. I've split all this logic up into the followiong

1. `client.go` This now just holds a client that's sole purpose is for making requests to Harness Saas
2. `map.go` This contains a custom map type that aggregates metrics by environment and exposes methods to add and retrieve metrics data from the map. The code for this used to exist in `client.go`, I've just copied and pasted it into another file and wrapped it in a custom type.
3.  `queue.go` This contains an in memory queue implementation where we have a method that lets us add metrics to the queue and a `Listener` method that returns a channel so we can listen for metrics being flushed every minute or when the max memory in MB is reached.
4. `worker.go` This creates a worker in the same way we do in the ff-metrics-service that's job is to listen for a events on a redis stream, add them to the in memory queue and then send them on to Saas.

**Why**

- If we only flush based on a timer of 1 minute if we receive a lot of large metrics requests the Primary can OOM

**Testing**

- Manual testing locally
- Tested this out in QA with a dev build and it's working as expected with metrics going from SDK -> Replica -> Redis -> Primary -> Saas
- Load testing this out in QA it resolves the OOM issue we were having
- I'm going to add unit tests for all these new types I've added, I just didn't want the PR to be any bigger
